analyzer/src/main/kotlin/managers/Cargo.kt

```diff
@@ -0,0 +1,290 @@
+        while (workingDir != analyzerRoot.parentFile) {
+            val lockfile = File(workingDir, "Cargo.lock")
+            if (lockfile.exists()) {
+                return lockfile
+            }
+            workingDir = workingDir.parentFile
+        }
+        // we reached the root directory we are analyzing
+        throw IllegalArgumentException("missing Cargo.lock file")
+    }
+
+    private fun readHashes(lockfile: File): Map<String, String> {
+        val contents = Toml().read(lockfile)
+        val metadata = contents.getTable("metadata") ?: return emptyMap()
+        val metadataMap = metadata.toMap()
+        return metadataMap.asSequence().mapNotNull {
+            val key = it.key
+            val value = it.value as? String
+            if (key != null && value != null) {
+                Pair(key, value)
+            } else {
+                null
+            }
+        }.toMap()
+    }
+
+    private fun resolveDependenciesTree(
+        metadata: JsonNode,
+        packages: Map<String, Package>,
+        filter: (pkgId: String, depId: String) -> Boolean = { _, _ -> true }
+    ): SortedSet<PackageReference> {
+        val resolve = metadata["resolve"]
+        val nodes = resolve["nodes"]
+        val root = resolve["root"].textValueOrEmpty()
+
+        return resolveDependenciesOf(
+            root, nodes, packages, filter
+        ).dependencies
+    }
+
+    private fun resolveDependenciesOf(
+        id: String,
+        nodes: JsonNode,
+        packages: Map<String, Package>,
+        filter: (pkgId: String, depId: String) -> Boolean
+    ): PackageReference {
+        val node = nodes.find { it["id"].textValueOrEmpty() == id }!!
+        val depsReferences = node["dependencies"].asSequence()
+            .map { it.textValue()!! }
+            .filter { filter(id, it) }
+            .map { resolveDependenciesOf(it, nodes, packages, filter) }
+        val pkg = packages[id]!!
+        val linkage = if (isProjectDependency(id)) PackageLinkage.PROJECT_STATIC else PackageLinkage.STATIC
+        return pkg.toReference(linkage, dependencies = depsReferences.toSortedSet())
+    }
+
+    private fun isDevDependencyOf(id: String, depId: String, metadata: JsonNode): Boolean {
+        val packages = metadata["packages"]
+        val pkg = packages.asSequence().find { it["id"].textValueOrEmpty() == id }!!
+        val depPkg = packages.asSequence().find { it["id"].textValueOrEmpty() == depId }!!
+        val dep = pkg["dependencies"].find {
+            val name = it["name"].textValueOrEmpty()
+            name == depPkg["name"].textValueOrEmpty()
+        }!!
+        return dep["kind"].textValueOrEmpty() == "dev"
+    }
+
+    // Check if a package is a project
+    //
+    // We treat all path dependencies inside of the analyzer root as project dependencies
+    private fun isProjectDependency(id: String): Boolean {
+        val pathRegex = Regex("""^.*\(path\+file://(.*)\)$""")
```

Comment:
This could go to some private constant so the Regex is not recompiled on each function call.

Reply:
Ping?

Reply:
Sorry, somehow I missed several comments.